// anvil.js/next/src/fileMachine.js
var fileMachineFactory = function( _, fp, scheduler, postal, machina ) {

	// ## FileMachine
	// This state machine manages the pipeline and interactions
	// for this file during the build.
	// The filePath is the fully qualified or relative path to
	// the file and the onComplete callback will receive a handle
	// to the created file state machine.
	var FileMachine = function( filePath, searchPatterns, replacePatterns, onCreated ) {
		fp.stat( filePath, function( metadata ) {
			var machine = new machina.Fsm( {
				broadcast: postal.channel( "fileEvents" ),
				dependents: 0,
				ext:  function() { return path.extname( this.name ); },
				fullPath: path.resolve( filePath ),
				imports: [],
				initialState: "initilization",
				lastModified: metadata.lastModified,
				name: path.basename( filePath ),
				relativePath: path.resolve( "./", filePath ),
				replacePatterns: replacePatterns,
				searchPatterns: searchPatterns,
				watcher: fs.watch( filePath, this.onFileChange ),
				workingPath: "",

				currentFullPath: function() {
					return fp.buildPath( [ this.workingPath, this.name ] );
				},
				
				onFileChange: function() {},
				
				broadcastStep: function( step ) {
					this.broadcast.publish( {
						topic: this.currentFullPath(),
						body: {
							file: this,
							event: step
						}
					} );
				},

				copy: function( to, onComplete ) {
					var self = this;
					fp.copy( [ this.workingPath, this.name ], to, function( newFull ) {
						self.workingPath = path.dirname( newFull );
						self.name = path.basename( newFull );
						onComplete();
					} );
				},

				findImports: function( onComplete ) {
					var self = this,
						sourcePath = this.currentFullPath(),
						imports = [],
						match;

					fp.read( this.currentFullPath(), function( content ) {
						_.each( self.searchPatterns, function( pattern ) {
							if( ( match = content.match( pattern ) ) != undefined ) {
								imports = imports.concat( match );
							}
						} );
						
						_.each( imports, function ( imported ) {
							var importName = imported.match( /['\"].*['\"]/ )[ 0 ].replace( /['\"]/g, "" );
							var importPath = path.resolve( currentFullPath, importName );
							if( path.existsSync( importPath ) ) {
								self.imports.push( importPath );
							}
						} );
						onComplete();
					} );
				},

				transform: function( type, newName, transform, onComplete ) {
					var self = this;
					fp.transform(
						[ this.workingPath, this.name ],
						transform,
						[ this.workingPath, newName ],
						function( err ) {
							if( !err ) {
								file.name = newFile;
								onComplete( file );
							} else {
								self.raiseEvent( {
									type: type,
									message: type + " failed for " + self.fullPath,
									error: err
								} );
								onComplete( file, err );
							}
						} );
				},

				states: {
					"initialization": {
						_onEnter: function() {
							_.bindAll( this );
							this.broadcastStep("ready");
						},
						"copy": function( command ) {
							this.copy( command.path, function() {
								this.transition( "scanning" );
							} );
						}
					},
					"scanning": {
						_onEnter: function() {

						},
						"imported": function() {
							this.dependents++;
						},
						"*": function() {

						}
					} 
				}
			} ) );
		} );

};

modules.exports = fileMachineFactory;
